environs/manual/linux/linux.go

```diff
@@ -10,54 +11,90 @@ import (
-uname -m
-grep MemTotal /proc/meminfo
-cat /proc/cpuinfo`
-
-// CheckProvisioned checks if any juju init service already
-// exist on the host machine.
-var CheckProvisioned = checkProvisioned
-
-func checkProvisioned(host string) (bool, error) {
-	logger.Infof("Checking if %s is already provisioned", host)
+// InitUbuntuUser adds the ubuntu user if it doesn't
+// already exist, updates its ~/.ssh/authorized_keys,
+// and enables passwordless sudo for it.
+//
+// InitUbuntuUser will initially attempt to login as
+// the ubuntu user, and verify that passwordless sudo
+// is enabled; only if this is false will there be an
+// attempt with the specified login.
+//
+// authorizedKeys may be empty, in which case the file
+// will be created and left empty.
+//
+// stdin and stdout will be used for remote sudo prompts,
+// if the ubuntu user must be created/updated.
+func InitUbuntuUser(host, login, authorizedKeys string, stdin io.Reader, stdout io.Writer) error {
+	logger.Infof("initialising %q, user %q", host, login)
 
-	script := service.ListServicesScript()
+	// To avoid unnecessary prompting for the specified login,
+	// initUbuntuUser will first attempt to ssh to the machine
+	// as "ubuntu" with password authentication disabled, and
+	// ensure that it can use sudo without a password.
+	//
+	// Note that we explicitly do not allocate a PTY, so we
+	// get a failure if sudo prompts.
+	cmd := ssh.Command("ubuntu@"+host, []string{"sudo", "-n", "true"}, nil)
+	if cmd.Run() == nil {
+		logger.Infof("ubuntu user is already initialised")
+		return nil
+	}
 
-	cmd := ssh.Command("ubuntu@"+host, []string{"/bin/bash"}, nil)
-	var stdout, stderr bytes.Buffer
-	cmd.Stdout = &stdout
+	// Failed to login as ubuntu (or passwordless sudo is not enabled).
+	// Use specified login, and execute the initUbuntuScript below.
+	if login != "" {
+		host = login + "@" + host
+	}
+	script := fmt.Sprintf(initUbuntuScript, utils.ShQuote(authorizedKeys))
+	var options ssh.Options
+	options.AllowPasswordAuthentication()
+	options.EnablePTY()
+	cmd = ssh.Command(host, []string{"sudo", "/bin/bash -c " + utils.ShQuote(script)}, &options)
+	var stderr bytes.Buffer
+	cmd.Stdin = stdin
+	cmd.Stdout = stdout // for sudo prompt
 	cmd.Stderr = &stderr
-	cmd.Stdin = strings.NewReader(script)
 	if err := cmd.Run(); err != nil {
 		if stderr.Len() != 0 {
 			err = fmt.Errorf("%v (%v)", err, strings.TrimSpace(stderr.String()))
 		}
-		return false, err
-	}
-
-	output := strings.TrimSpace(stdout.String())
-	provisioned := strings.Contains(output, "juju")
-	if provisioned {
-		logger.Infof("%s is already provisioned [%q]", host, output)
-	} else {
-		logger.Infof("%s is not provisioned", host)
+		return err
 	}
-	return provisioned, nil
+	return nil
 }
 
+const initUbuntuScript = `
+set -e
+(id ubuntu &> /dev/null) || useradd -m ubuntu -s /bin/bash
+umask 0077
+temp=$(mktemp)
+echo 'ubuntu ALL=(ALL) NOPASSWD:ALL' > $temp
+install -m 0440 $temp /etc/sudoers.d/90-juju-ubuntu
```

Comment:
Even though this is all Linux, this needs be configurable. Pass in a `configPath` variable that is derived from our `github.com/juju/juju/juju/paths` package up the stack.